### PR TITLE
Scope build args to a single stage in a multi-stage build

### LINKED ISSUE #2210: buildah shows build-arg values in final image of multi-stage build
**Description**

When doing multi-stage builds, Buildah shows the build-arg -variable values it has received in the final image. This is not the behaviour docker build has. The use case was passing build time credentials from Jenkins to buildah and hiding them from the final image, so this is undesired.

I simplified the dockerfile to minimum to demonstrate the issue. Overwriting the value with ENV in builder -stage does not change the output in second stage, so I guess this is probably due to build-arg -handling.

Dockerfile used:

```
from alpine:latest as builder

ARG SECRET

RUN touch /tmp/file

from alpine:latest

COPY --from=builder /tmp/file /tmp/file

RUN apk update
```
Docker build command
`docker build -t simple-docker --build-arg SECRET="kissa123" .`

Buildah build command
`buildah build-using-dockerfile  -t simple-buildah --build-arg SECRET="kissa123" .`

**Describe the results you received:**
```
[vagrant@localhost simple]$ podman history --no-trunc localhost/simple-buildah --format="{{.CreatedBy }}"
|1 SECRET=kissa123 /bin/sh -c apk update
/bin/sh -c #(nop) COPY file:baccd90931c4c200e3cb7453974fe6e01dbe45e47b6754b8a540d6eca976cfba in /tmp/file
/bin/sh -c #(nop) CMD ["/bin/sh"]
/bin/sh -c #(nop) ADD file:e69d441d729412d24675dcd33e04580885df99981cec43de8c9b24015313ff8e in /

```

**Describe the results you expected:**
```
[vagrant@localhost simple]$ docker history --no-trunc simple-docker --format="{{.CreatedBy }}"
/bin/sh -c apk update
/bin/sh -c #(nop) COPY file:e29d5c9f00a7d18d5705c28f95d7aaa320cc846ec833f9ab00036f292a30f637 in /tmp/file 
/bin/sh -c #(nop)  CMD ["/bin/sh"]
/bin/sh -c #(nop) ADD file:e69d441d729412d24675dcd33e04580885df99981cec43de8c9b24015313ff8e in / 
```

**Output of `rpm -q buildah`**

```
buildah-1.14.0-1.1.el8.x86_64
```

**Output of `buildah version`:**

```
Version:         1.14.0
Go Version:      go1.12.12
Image Spec:      1.0.1-dev
Runtime Spec:    1.0.1-dev
CNI Spec:        0.4.0
libcni Version:  
image Version:   5.2.0
Git Commit:      
Built:           Thu Jan  1 00:00:00 1970
OS/Arch:         linux/amd64

```

**Output of `podman version` if reporting a `podman build` issue:**

```
Version:            1.8.0
RemoteAPI Version:  1
Go Version:         go1.12.12
OS/Arch:            linux/amd64

```

**Output of `cat /etc/*release`:**

```
NAME="CentOS Linux"
VERSION="8 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="8"
PLATFORM_ID="platform:el8"
PRETTY_NAME="CentOS Linux 8 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:8"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-8"
CENTOS_MANTISBT_PROJECT_VERSION="8"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="8"

CentOS Linux release 8.1.1911 (Core) 
CentOS Linux release 8.1.1911 (Core) 

```

**Output of `uname -a`:**

```
Linux localhost.localdomain 4.18.0-147.5.1.el8_1.x86_64 #1 SMP Wed Feb 5 02:00:39 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux

```

**Output of `cat /etc/containers/storage.conf`:**

Omitting all comments, this file has not been changed from defaults
```
[storage]

driver = "overlay"
runroot = "/var/run/containers/storage"
graphroot = "/var/lib/containers/storage"

[storage.options]
additionalimagestores = [
]
size = ""
override_kernel_check = "true"

[storage.options.thinpool]
ostree_repo = ""

skip_mount_home = "false"
```

**Output of `docker --version` for reproducing**
`Docker version 19.03.7, build 7141c199a2`

### PR DESCRIPTION
**Pull request by carbonin**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

This PR uses the changes made in https://github.com/openshift/imagebuilder/pull/151 to handle arguments the same way `docker build` does. In particular, it scopes arguments to the stage in which they are defined and only records arguments in a layer's history if they could have been used in that layer.

#### How to verify it
```
$ cat Dockerfile 
FROM alpine
ARG THING

FROM alpine
RUN echo "$THING" > things
        
$ buildah bud --layers --build-arg THING=things
STEP 1: FROM alpine
STEP 2: ARG THING
--> Using cache 232af6ca4a94e52dbef13f6da08c62b4172eaff7ee2e93cab08aceb4b00e6f81
STEP 3: FROM alpine
STEP 4: RUN echo "$THING" > things
--> Using cache fad9788d65a3062cc823516c8fff73b39e914463c709149a2855cbea61a10abe
fad9788d65a3062cc823516c8fff73b39e914463c709149a2855cbea61a10abe
$ podman run --rm fad9788d65a3062cc823516c8fff73b39e914463c709149a2855cbea61a10abe cat things
things
```

The above `podman run` command should return an empty file

#### Which issue(s) this PR fixes:

Fixes #2210

#### Does this PR introduce a user-facing change?

Yes, it significantly changes the way arguments behave. While it's not a change to how `buildah` would be used. Images built with the same Dockerfile before and after this change could be very different.

```release-note
Altered the behavior of the `--build-arg` flag and `ARG` commands to mirror `docker build`.
In particular, the following behaviors have changed:

- An ARG is only available after its ARG command in the current stage.
    - Previously, anything provided using the --build-arg flag could be accessed in any stage. After this change, accessing a build arg provided on the command line will require a corresponding `ARG` command in the stage before it is accessed.
    - Additionally, "heading" args (ARG commands before the first FROM) also now require an additional ARG declaration in the stage to be accessed. Previously, they were accessible without the additional ARG command.

- A later ARG default value should override an earlier one in the same stage

FROM alpine
ARG FOO=foo
ARG FOO=bar
RUN echo "$FOO"

The above Dockerfile should print "bar". Previously, the behavior was the opposite, an arg was not changed once set.

Generally this makes buildah handle args as described in https://docs.docker.com/engine/reference/builder/#arg
```

### CODE REVIEW

**Comment by rhatdan** [comment]
You will need to open a PR in the openshift/imagebuilder  repo

**Comment by rhatdan** [comment]
@carbonin The mult-stage build args tests are failing.  Is that a problem with your PR or the tests?

**Comment by carbonin** [comment]
I think it's a problem with my inexperience with GO.

I'm not sure how to correctly change a dependency from one github org to another.
I just realized that the dependencies are fully qualified in the `import` section. But when I change those and run `go mod vendor` the build gives me these errors:

```
GO111MODULE=on go build -mod=vendor -ldflags '-X main.GitCommit=224a69a0 -X main.buildInfo=1584710508 -X main.cniVersion=v0.7.2-0.20190904153231-83439463f784'  -o buildah -tags "seccomp selinux    " ./cmd/buildah
# github.com/containers/buildah/imagebuildah
imagebuildah/build.go:260:26: cannot use mainNode (type *"github.com/openshift/imagebuilder/dockerfile/parser".Node) as type *"github.com/carbonin/imagebuilder/dockerfile/parser".Node in argument to NewExecutor
imagebuildah/stage_executor.go:819:72: cannot use stage.Node (type *"github.com/openshift/imagebuilder/dockerfile/parser".Node) as type *"github.com/carbonin/imagebuilder/dockerfile/parser".Node in argument to s.executor.getCreatedBy
imagebuildah/stage_executor.go:886:52: cannot use *"github.com/carbonin/imagebuilder/dockerfile/parser".Node literal (type *"github.com/carbonin/imagebuilder/dockerfile/parser".Node) as type *"github.com/openshift/imagebuilder/dockerfile/parser".Node in argument to ib.RequiresStart
imagebuildah/stage_executor.go:886:53: cannot use children[i + 1:] (type []*"github.com/openshift/imagebuilder/dockerfile/parser".Node) as type []*"github.com/carbonin/imagebuilder/dockerfile/parser".Node in field value
imagebuildah/stage_executor.go:898:55: cannot use node (type *"github.com/openshift/imagebuilder/dockerfile/parser".Node) as type *"github.com/carbonin/imagebuilder/dockerfile/parser".Node in argument to s.digestSpecifiedContent
imagebuildah/stage_executor.go:908:67: cannot use node (type *"github.com/openshift/imagebuilder/dockerfile/parser".Node) as type *"github.com/carbonin/imagebuilder/dockerfile/parser".Node in argument to s.executor.getCreatedBy
imagebuildah/stage_executor.go:917:65: cannot use node (type *"github.com/openshift/imagebuilder/dockerfile/parser".Node) as type *"github.com/carbonin/imagebuilder/dockerfile/parser".Node in argument to s.executor.getCreatedBy
imagebuildah/stage_executor.go:947:55: cannot use node (type *"github.com/openshift/imagebuilder/dockerfile/parser".Node) as type *"github.com/carbonin/imagebuilder/dockerfile/parser".Node in argument to s.digestSpecifiedContent
imagebuildah/stage_executor.go:951:44: cannot use node (type *"github.com/openshift/imagebuilder/dockerfile/parser".Node) as type *"github.com/carbonin/imagebuilder/dockerfile/parser".Node in argument to s.intermediateImageExists
imagebuildah/stage_executor.go:1005:55: cannot use node (type *"github.com/openshift/imagebuilder/dockerfile/parser".Node) as type *"github.com/carbonin/imagebuilder/dockerfile/parser".Node in argument to s.digestSpecifiedContent
imagebuildah/stage_executor.go:1005:55: too many errors
make: *** [Makefile:50: binary] Error 2
```

I'm not sure how to deal with the types being different between my fork and the main package

**Comment by rhatdan** [comment]
Change go.mod and then run
make vendor-in-container
Which should pull in the latest vendor code.

**Comment by TomSweeneyRedHat** [comment]
@carbonin what @rhatdan said.  If you're ever looking for the vendor instructions in the future, we keep them at the bottom of the install.md page:  https://github.com/containers/buildah/blob/master/install.md#vendoring

**Comment by TomSweeneyRedHat** [comment]
@carbonin what @rhatdan said.  If you're ever looking for the vendor instructions in the future, we keep them at the bottom of the install.md page:  https://github.com/containers/buildah/blob/master/install.md#vendoring

**Comment by carbonin** [comment]
Thanks @TomSweeneyRedHat https://github.com/containers/buildah/blob/master/install.md#vendor-from-your-own-fork was exactly what I needed.

It looks like there is some issue with `buildah` causing that test to fail.
Using this Dockerfile:
```Dockerfile
FROM alpine
ARG SECRET
RUN echo $SECRET

FROM alpine
RUN echo "$SECRET"
```

`imagebuilder` + my branch:
$ ../imagebuilder --build-arg SECRET=secret .
--> FROM alpine as 0
--> ARG SECRET
--> RUN echo $SECRET
secret
--> FROM alpine as 1
--> RUN echo "$SECRET"

--> Committing changes ...
--> Done

`buildah` + the vendor changes in my previous commit:
```
$ ../buildah bud --build-arg SECRET=secret
STEP 1: FROM alpine
STEP 2: ARG SECRET
STEP 3: RUN echo $SECRET
secret
STEP 4: FROM alpine
STEP 5: RUN echo "$SECRET"
secret
STEP 6: COMMIT
Getting image source signatures
Copying blob 5216338b40a7 skipped: already exists  
Copying blob ab4c80a1663f done  
Copying config 30e353c1ef done  
Writing manifest to image destination
Storing signatures
--> 30e353c1efd
30e353c1efd0612b52cb4f12d53a825d5888e20499015207b220dc4dcd78e936
```

Looking in to this now.

**Comment by carbonin** [comment]
I'm fairly sure I have this last problem narrowed down to https://github.com/containers/buildah/blob/aba0d4dfe1456433b6463062db02de122a2b8fe3/imagebuildah/executor.go#L303-L307, but I'm still trying to work out a way for the history to be accurate in a multi-stage situation.

Right now, https://github.com/containers/buildah/blob/aba0d4dfe1456433b6463062db02de122a2b8fe3/imagebuildah/executor.go#L211-L229 traverses the whole Dockerfile looking for `ARG` commands matching what was provided in the command line. The problem is that it isn't multi-stage aware so build args are "leaking" into later stages even if they are not used there.

**Comment by carbonin** [comment]
@TomSweeneyRedHat @rhatdan I think this one is ready for review. I'll leave it as a work in progress until https://github.com/openshift/imagebuilder/pull/151 is merged though as it still has the commit to pull in my branch.

**Comment by rhatdan** [comment]
You should remove the [wip] flag

**Comment by rhatdan** [comment]
Can you merge your RPS together?

**Comment by rhatdan** [comment]
@nalind @bparees PTAL
Want to make sure this does not break openshift/builder as well.

**Comment by carbonin** [comment]
Removed [WIP]. Should I rebase out the commit that points to my `imagebuilder` branch now? That will make the tests go red on this PR.

> Can you merge your PRs together?

I'm not sure what you mean by this. I think they should be merged at the same time, but I didn't think I should actually copy over the patch to `imagebuilder` to here. I thought this one would wait until the change is merged to `imagebuilder` then I would update the dependency here and remove my branch reference.

**Comment by rhatdan** [comment]
I meant squash them together.

**Comment by rhatdan** [comment]
@carbonin Please vendor in the newest the openshift/imagebuilder and then we can review this PR.

**Comment by carbonin** [comment]
:+1: Updated imagebuilder to `master` and added some more detail to the PR description.

**Comment by rhatdan** [comment]
LGTM
@nalind @TomSweeneyRedHat @QiWang19 @vrothberg PTAL

File: go.mod [hunk 1]
```diff
@@ -27,7 +27,7 @@ require (
 	github.com/opencontainers/runtime-tools v0.9.0
 	github.com/opencontainers/selinux v1.4.0
 	github.com/openshift/api v0.0.0-20200106203948-7ab22a2c8316
-	github.com/openshift/imagebuilder v1.1.3
+	github.com/openshift/imagebuilder v1.1.4-0.20200325224900-8de717452756
```
Comment (TomSweeneyRedHat) [inline go.mod]: I'd prefer having a full imagebuilder release here if we can.  I know @carbonin can't make that happen from that end.

Comment (rhatdan) [reply go.mod]: Your wish is my command.
https://github.com/openshift/imagebuilder/releases/tag/v1.1.4

**Comment by TomSweeneyRedHat** [comment]
Looks OK to me, but I'd really like a deep dive from @nalind as he just finished diving deeply on the imagebuilder side of this.

Please don't merge this before 1.4.5 drops later today unless we update the imagebuilder to 1.4 rather than a dev.

**Comment by carbonin** [comment]
Thanks for your help and patience on this @rhatdan @TomSweeneyRedHat

**Comment by rhatdan** [comment]
Thank you for the PR and actually fixing the issue.

**Pull request merged**